### DISCUSSION TITLE
Integration with Paperless NGX

### DISCUSSION THREAD

**Discussion by bmschwa** [firefly-iii/firefly-iii#9467] [General]
A widely used document management system in the self-hosted world is [Paperless NGX](https://github.com/paperless-ngx/paperless-ngx).   As an enhancement, what about an integration with the [Paperless API](https://docs.paperless-ngx.com/api/) so that documents that are important enough to save as financial records get stored in the same place as all one's other self hosted docs? 

The way I'd want the integration to behave is that if you have a paperless doc server configured, rather than use the filesystem to store files, you would instead use the paperless api to save docs.  

... if there's interest, and its not going to be nixed by @JC5, I'd put some time into building out the integration.

**Comment by Heart1010** [comment]
I always wanted to test Paperlese NGX but never found the time for it (using Joplin for jow)... this feature could be really handy and a reason more to finally install Paperless 😉

**Comment by JC5** [comment]
Your best bet is to make a file system driver for Laravel, but I got to be honest the interest seems a little low.

**Comment by molnarti** [comment]
transactions  have an external URL field. what i do is to link the paperless URL to the transactions where i want to keep that info

**Comment by joeshmoe57** [comment]
I have a webhook do a search on my Paperless instance for relevant receipts and populate the External URL field @molnarti mentioned. I also populate some Paperless fields from the Firefly transaction (cross-link, amount). Works great

**Reply by molnarti** [reply]
hey, mind to share your webhook? sounds cool and would be very happy to try it

**Reply by joeshmoe57** [reply]
I can try to do that... it's built into some other code I have, so I'll need to piece it out for sharing (I use 1 "big" webhook that does the paperless stuff, classification, and Amazon lookups, but it's very much personalized at the moment).

Generally, it's pretty straightforward Python code that uses the requests library to hit the [Paperless Search API](https://docs.paperless-ngx.com/api/#searching-for-documents) and the document update API if you're interested in updating on the Paperless side (full schema is viewable at <your-paperless-insance>/api/schema/view)

**Reply by Triserden** [reply]
I honestly completely forgot about webhooks, very interesting use case. I also have a Paperless instance, and I might make a standalone app for this when I have time. How did you do the searching? Just "Amount Due" + amount on the invoice tag/category?

**Reply by joeshmoe57** [reply]
I'm sure there are many ways to succeed here, and individuals' setups of both Paperless and FF will dictate what makes sense.

Personally, my search is primarily merchant and date based, goal being to find a single receipt so it can be linked. Finding multiple is less helpful (but I add those details to the Notes field so I can manually disambiguate later)
- I search for `type:receipt`
    - my Paperless has gotten pretty good at IDing these, so this is mostly 0 effort, but I generally populate during Paperless consume
- I include the relevant account name
    - mostly "destination_name" from FF, but if it's a deposit, I use the "source_name" instead
    - Paperless has also gotten good at doing this without effort for everything but the one-offs
    - this will hit on the "Correspondent" field, but also hit if Paperless OCRs a one-off from the receipt (which it often does)
- I search a date range from the date in the FF transaction and up to 2 days prior (may need some more playing; I often find transactions resolve (FF side) a day or two after the actual purchase (Paperless side), but you don't want too big a window
   - syntax was a bit finicky, I use something like `created:[20250930 to 20251002]`
- I do a small tag lookup/conversion for some frequent items like gas that often create collisions
   - i.e. if the FF transaction has category "Transportation - Gas", add `tag:Gas` to the Paperless search
   - helpful because I can't disambiguate automatically, so this helps a couple common cases
- sometimes I still get multiple hits, so I recently added including the amount in the search
   - I actually only do this if I actually have multiple hits (via a 2nd search) 
   - I found including it from the start can be too restrictive if Paperless doesn't pick up the amount correctly. Again, may vary person to person a bit

So at the end of the day, my `query` sent to Paperless looks something like `type:receipt created:[20250901 to 20250902] Target`

**Reply by Triserden** [reply]
That's interesting, receipts? Do you scan those paper receipts you get everytime? Slightly offtopic but I'm very interested in how you did that.

Thanks for the details, I'll certainly keep it in mind.

**Reply by joeshmoe57** [reply]
I tend to scan receipts using the [Paperless Mobile](https://github.com/astubenbord/paperless-mobile) app, but other sources include picking them up from Email or dropping the PDF into the consume directory.

What kind of documents were you looking to link to FF if not receipts?

**Reply by Triserden** [reply]
Mostly invoices from emails and the like.

**Comment by nottheend** [comment]
This is interesting - I am working on some script to extract item details e.g. from supermarket receipts. The aim is to have a better understanding for which category I am spending money (vegetables, protein, sweets...).
In the beginning I wanted to create via API one subtransation per item. Summarising one receipt in one grouped transaction. But this seemed to be too complicated per API for the beginning, so I am only categorising items separately for now and enter the receipt as a sum to firefly-iii
